### ISSUE TITLE
Safari forOf and spread issues for Set object.

### ISSUE THREAD

**Issue by patrick-martin-shoutpoint** [google/traceur-compiler#1810]
When using the Set object, the following code fails in Safari 8 desktop and mobile:
"Error: undefined is not a function (evaluating 'f.next()')"
```
var set = new Set([1,2,3,4]);

var arr = [...set]; // FAIL

for (var v of set) // FAIL
  console.log(v);

var vals = set.values();
for (var v of vals) // FAIL
  console.log(v);
```
Note: This code fails to compile in the Traceur REPL on Safari.
https://google.github.io/traceur-compiler/demo/repl.html#var%20set%20%3D%20new%20Set(%5B1%2C2%2C3%2C4%5D)%3B%0A%0Avar%20arr%20%3D%20%5B...set%5D%3B%0A%0Afor%20(var%20v%20of%20set)%0A%20%20console.log(v)%3B%0A%0Avar%20vals%20%3D%20set.values()%3B%0Afor%20(var%20v%20of%20vals)%0A%20%20console.log(v)%3B%0A
```
spread@https://google.github.io/traceur-compiler/bin/traceur.js:1316:38

getAnonymousModule@https://google.github.io/traceur-compiler/bin/traceur.js:681:34
eval code
eval@[native code]
evaluateCodeUnit@https://google.github.io/traceur-compiler/bin/traceur.js:29800:36
evaluate@https://google.github.io/traceur-compiler/bin/traceur.js:29956:50
evaluate@https://google.github.io/traceur-compiler/bin/traceur.js:30257:37
handleCodeUnitLoaded@https://google.github.io/traceur-compiler/bin/traceur.js:30170:24
module@https://google.github.io/traceur-compiler/bin/traceur.js:30056:32
module@https://google.github.io/traceur-compiler/bin/traceur.js:30317:41
transcode
compileContents
compile
wrappedFunc
```

**Comment by caitp** [comment]
yeah, iterator protocol in JSC is still a work in progress. last I checked (in jan/feb), @@iterator wasn't exposed at all in JS, and was only implemented for specific types (just arrays, maybe).

maybe it would make sense to always use the polyfilled version for JSC?

**Comment by arv** [comment]
We should probably be a lot stricter when it comes to determining when to use the polyfills. One suggestion is to use the polyfills if there is no `[Symbol.iterator]`.

**Comment by cwalther** [comment]
I just ran into a similar issue with Safari on iOS 8.1.3. The problematic cases in my case were for-of over an Array as well as construction of a Map from an Array (or iterable). The specific reasons were that
- Safari natively implements `Array.prototype.entries` and `Array.prototype.keys`, but the iterators returned by them lack a `next` method.
- Safari natively implements `Map`, but the constructor does not take an argument.

I can work around the problems by including the following as the first script of the page:
````JavaScript
// Remove insufficient built-in implementations so that the polyfills from traceur-runtime kick in.
if (Array.prototype.entries && typeof [1].entries().next !== 'function') {
	delete Array.prototype.entries;
}
if (Array.prototype.keys && typeof [1].keys().next !== 'function') {
	delete Array.prototype.keys;
}
try {
	new Map([]);
}
catch(e) {
	delete Map;
}
````

I assume these checks could be built into the [polyfillArray](https://github.com/google/traceur-compiler/blob/master/src/runtime/polyfills/Array.js#L148) and [polyfillMap](https://github.com/google/traceur-compiler/blob/master/src/runtime/polyfills/Map.js#L195) functions but have not tried that.

**Comment by arv** [comment]
> Safari natively implements Array.prototype.entries and Array.prototype.keys, but the iterators returned by them lack a next method.

WHAT!! How do you iterate over them then? How can they ship something so broken... makes me very sad.

@cwalther What does Map.length return in Safari? If it says 0 we should be able to use that instead of using the try/catch.

**Comment by caitp** [comment]
I believe they've since fixed this in JSC (but still don't expose Symbol.iterator, afaik)

**Comment by cwalther** [comment]
I have no idea what the purpose of these iterators is, they appear to have no methods (or other properties) at all besides the ones inherited from Object.

`Map.length` indeed returns 0. (I wasn’t even aware that functions had a length property, thanks for pointing me to that!)

Interestingly, `Set.length` is 0 as well, but `new Set([4, 5])` still works (`Set([4, 5])` on the other hand throws `TypeError: Set does not accept arguments when called as a function `).

**Comment by arv** [comment]
> TypeError: Set does not accept arguments when called as a function

Sad. So misleading. Set throws when called. 

https://people.mozilla.org/~jorendorff/es6-draft.html#sec-set-iterable

Step 1

**Comment by cwalther** [comment]
I just see in that document that [`Map.length` *should* be 0](https://people.mozilla.org/~jorendorff/es6-draft.html#sec-properties-of-the-map-constructor), unless I’m misunderstanding something, so maybe checking for that is not a suitable way of detecting insufficient implementations after all.

By the way, [es6-shim](https://github.com/paulmillr/es6-shim) also fixes the issues I had on that version of Safari, and it has [a whole bunch of criteria](https://github.com/paulmillr/es6-shim/blob/master/es6-shim.js#L2521) for checking whether to replace Map, maybe that’s a more reliable source. (I didn’t use es6-shim as a final solution because I was wary of potential unintended interactions between es6-shim and traceur-runtime.)

**Comment by arv** [comment]
You are correct. Map.length should be 0.

I think a suitable test is to see if Map.prototype[Symbol.iterator] is a function.

**Comment by zloirock** [comment]
@cwalther @arv bad idea. `es6-shim` also has a problem - it's replace collections in all modern environment. Maybe it's partly fixed after yesterday's reworking. My point about collections polyfilling [here](https://github.com/paulmillr/es6-shim/issues/326#issuecomment-90737758).
